textalyze.rb

```diff
@@ -1,66 +1,101 @@
-#
-# Returns: A hash where every item is a key whose value is the number of times
-#          that item appears in the array
-#
-# Prints:  Nothing
+if ARGV.size == 0
+  puts "You didn't tell us what file to read. Try this commmand:"
+  puts
+  puts "  ruby textalyze.rb random_file.txt"
+  exit # Exits the program
+end
 
-# Here are some examples:
-#     item_counts(["I", "am", "Sam", "I", "am"])
-#       # => {"I" => 2, "am" => 2, "Sam" => 1}
-#
-#     item_counts([10, 20, 10, 20, 20])
-#       # => {10 => 2, 20 => 3}
-#
-# In short, item_counts(array) tells us how many times each item appears
-# in the input array.
+file_name = ARGV[0] #first argument
+file_contents = File.read(file_name) # Files will have to be in same directory
+# as textalyze.rb - simimlar to v0.4.
 
 def item_counts(array)
   counts = {} # Initialize counts to an empty Hash
+  array.each do |key|
+    if counts[key] # If this key exists
+      counts[key] += 1 # Increment the key
+    else # If this key doesn't exist
+      counts[key] = 1 # Create the key, set value to 1
+    end
+  end
+  # print_counts(counts)
+  counts
+end
 
-  array.each do |item|
-    # Add code here to modify the "counts" hash accordingly
-    # You'll need to handle two cases:
-    #   1. The first time we've seen a particular item in the array
-    #   2. The second-or-later time we've seen a particular item in the array
+def print_counts(counts_hash)
+  puts
+  counts_hash.each do |key, value|
+    puts "#{key}   #{value}"
   end
+  puts
+end
+
+def string_to_array(string) # returns array of all characters in the string
+  sanitize(string)
+  string.split(//)
+end
+
+def sanitize(string)
+  string.downcase!
+  string.gsub!(/\W+/, '') # Remove all non-word chars
+end
 
-  counts # This returns the "counts" hash
+# v1.1
+def frequency(string)
+  hash = item_counts(string_to_array(string)) # returns counts hash
+  hash.each do |key, value|
+    hash[key] = value.to_f/string.size.to_f
+  end
+  hash
 end
 
-# "p" prints something to the screen in a way that's friendlier
-# for debugging purposes than print or puts.
+# Notes on v.1.2
+# remove all non word chars - add to sanitize
+# take hash as input
+# round hash values to 2 places
+# force 0.00 format, figure our alignment
+# format as percent in output
+# scale bars to terminal width
+# get the bars to print out correctly next to the values
 
-p item_counts([1,2,1,2,1]) == {1 => 3, 2 => 2}
-p item_counts(["a","b","a","b","a","ZZZ"]) == {"a" => 3, "b" => 2, "ZZZ" => 1}
-p item_counts([]) == {}
-p item_counts(["hi", "hi", "hi"]) == {"hi" => 3}
-p item_counts([true, nil, "dinosaur"]) == {true => 1, nil => 1, "dinosaur" => 1}
-p item_counts(["a","a","A","A"]) == {"a" => 2, "A" => 2}
+def print_histo(stats)
+  ordered_stats = stats.sort
+  print "Calculating letter frequencies...done"
+  ordered_stats.each do |key, value|
+    percent = (value*100).round(2)
+    puts
+    print "[#{key}] [#{'%5.2f' % percent}%] " # This was trial and error.
```

Comment:
Oh, and actually, this is the key to calculating that number `13` dynamically.  Using `sprintf`, we get the `String` that will be printed at the start of each line.  Assuming they're all the same length, we can do...

```ruby
key_data_string = sprintf("[%s] [%5.2f%%]", key, percent)
key_data_string.length # => 13
```

You'll have to organize your code in a way where you can pass `key_data_string.length` around so that it is eventually passed to your `terminal_width` method that now accepts an "offset" argument. :)